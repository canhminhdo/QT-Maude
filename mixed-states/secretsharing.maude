in quantum-circuits.maude

mod SECRETE-SHARING is
    pr CIRCUIT .

    vars C : Complex .

    ops a b : -> Complex .
    eq Pow(Abs(a), 2) .+ Pow(Abs(b), 2) = 1 .
    eq C .* Pow(Abs(a), 2) .+ C .* Pow(Abs(b), 2) = C .

    op PSI : -> Vector .
    eq PSI = a . |0> + b . |1> .
    
    op ES : -> Ensemble .
    eq ES = {(a . |0> + b . |1>) (x) |0>  (x) |0> (x) |0>, 1} .

    op init : -> Config .
    eq init = {
        (#qbits: findN(ES))
        (mState: convert(ES))
        (actions: (H(1) CX(1, 2) CX(1, 3) CX(0, 1) H(0) H(2) M(0) M(1) M(2) (c[1] == 1 ? X(3)) (c[0] == 1 ? Z(3)) (c[2] == 1 ? Z(3))))
        (prob: 1)
        (bits: empty)
        (isEnd: false)
    } .
endm

load model-checker .

mod SECRETE-SHARING-PREDS is
    pr SECRETE-SHARING .
    inc SATISFACTION .

    subsort Config < State .

    var OCs : Soup{OComp} .
    var PROP : Prop .
    var MS : MixedState .
    var Prob : Complex .
    var N : Nat .

    ***************************************************************************************
    *** Factor \gamma on a state for which |\gamma| = 1 is regarded as a 'global phase' ***
    *** States that differ only by a global phase are physically indistinguishable and  ***
    *** equivalent in all ways that are physically equivalent
    ***************************************************************************************
    op isSuccess : -> Prop .
    eq {(isEnd: true) (mState: MS) (prob: Prob) (#qbits: N) OCs} |= isSuccess
    = Prob > 0 implies
    tr[2](tr[1]((tr[0](MS, N)), N), N) == (I (x) I (x) I (x) (PSI x (PSI)^+)) 
    [print Prob] .
    eq {OCs} |= PROP = false [owise] .
endm

mod SECRETE-SHARING-CHECK is
    inc SECRETE-SHARING-PREDS .
    inc MODEL-CHECKER .
    inc LTL-SIMPLIFIER .

    op ssProp : -> Formula .
    eq ssProp = <> isSuccess .
endm

in ../debug.maude

red modelCheck(init, ssProp) .

--- search in SECRETE-SHARING : init =>* {(mState: MS:MixedState) (isEnd: true) (prob: P:Complex) OCs:Soup{OComp}}
--- such that P:Complex > 0 implies (qubitAt(Q:Qubit, 3) == qubitAt(qstate(init), 0) or (-1) . qubitAt(Q:Qubit, 3) == qubitAt(qstate(init), 0)) .

--- red (initQState).H(1) .
--- red (initQState).H(1).CX(1, 2) .
--- red (initQState).H(1).CX(1, 2).CX(1, 3) .
--- red (initQState).H(1).CX(1, 2).CX(1, 3).CX(0, 1) .
--- red (initQState).H(1).CX(1, 2).CX(1, 3).CX(0, 1).H(0) .