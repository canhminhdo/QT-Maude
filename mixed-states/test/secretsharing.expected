Reading in file: "symbolic-reasoning.maude"
Reading in file: "core.maude"
==========================================
fmod COMPLEX
==========================================
fmod DIRAC
==========================================
fmod MATRIX
Done reading in file: "core.maude"
==========================================
fmod STATES-AND-PROB
==========================================
view State&Prob
==========================================
fmod ENSEMBLE
==========================================
fmod MIXED-STATES
==========================================
fmod BIT
==========================================
fmod ACTION
==========================================
view Action
==========================================
fmod ACTION-LIST
==========================================
fmod QUANTUM-GATES
==========================================
fmod SYMBOLIC-REASONING
Done reading in file: "symbolic-reasoning.maude"
==========================================
fmod SOUP
==========================================
fmod OCOMP
==========================================
view OComp
==========================================
fmod CONFIG
==========================================
mod CIRCUIT
==========================================
reduce in SECRETE-SHARING-CHECK : modelCheck(init, ssProp) .
result Bool: true
Bye.
